### DISCUSSION TITLE
Questions on script data storage

### DISCUSSION THREAD

**Discussion by TimidScript** [violentmonkey/violentmonkey#1800] [Q&A] [answered]
Just a few questions for those who may know how it works.

1. Is script data stored in external blobs or is it in a database?
2. If I add data (GM_setValue) does it just write the whole blob again, even for a tiny change or is it like a database?
3. Baisically I want to know what happens when I use GM_setValue. I have a few scripts that store several MBs of data and would like to know what happens when I add a value to the data.
4. "Sync" seems to only store scripts and not their data. Is that correct. I have tried it on several browsers.

Thank you for any help

**Comment by tophf** [comment, marked as answer]
1. We use `chrome.storage.local`. The physical shape of storage depends on the browser.
2. The entire script's data is stored as one key.
3. It depends on the extension. In Violentmonkey the entire script's data is written every time. Newer versions of Violentmonkey throttle actual writing, but other extensions probably don't. In the future we may store each script's value separately.
4. Yes, #48.

**Reply by TimidScript** [reply]
> The physical shape of storage depends on the browser.

I have firefox.

>     2. The entire script's data is stored as one key.

Currently I have scripts that do dozens of wrties on load, and each GM_setValue/Delete rewrites whole blob. So it seems I need to store all script data as one object and just write once.

> In the future we may store each script's value separately.

I hope that is the case. Thank you very much for the help.
